sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/CollationTypeConstraints.scala

```diff
@@ -0,0 +1,89 @@
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.expressions
+
+import org.apache.spark.SparkException
+import org.apache.spark.sql.catalyst.analysis.TypeCheckResult
+import org.apache.spark.sql.catalyst.analysis.TypeCheckResult.DataTypeMismatch
+import org.apache.spark.sql.catalyst.util.CollationFactory
+import org.apache.spark.sql.types.{DataType, StringType}
+
+object CollationTypeConstraints {
+  def checkCollationCompatibility(
+      superCheck: => TypeCheckResult,
+      collationId: Int,
+      rightDataType: DataType): TypeCheckResult = {
+    val checkResult = superCheck
+    if (checkResult.isFailure) return checkResult
+    // Additional check needed for collation compatibility
+    val rightCollationId: Int = rightDataType.asInstanceOf[StringType].collationId
+    if (collationId != rightCollationId) {
+      return DataTypeMismatch(
+        errorSubClass = "COLLATION_MISMATCH",
+        messageParameters = Map(
+          "collationNameLeft" -> CollationFactory.fetchCollation(collationId).collationName,
+          "collationNameRight" -> CollationFactory.fetchCollation(rightCollationId).collationName
+        )
+      )
+    }
+    TypeCheckResult.TypeCheckSuccess
+  }
+
+  object CollationSupportLevel extends Enumeration {
+    type CollationSupportLevel = Value
+
+    val SUPPORT_BINARY_ONLY: Value = Value(0)
+    val SUPPORT_LOWERCASE: Value = Value(1)
+    val SUPPORT_ALL_COLLATIONS: Value = Value(2)
+  }
+
+  def checkCollationSupport(
+      superCheck: => TypeCheckResult,
+      collationId: Int,
+      functionName: String,
+      collationSupportLevel: CollationSupportLevel.CollationSupportLevel)
+  : TypeCheckResult = {
+    val checkResult = superCheck
+    if (checkResult.isFailure) return checkResult
+    // Additional check needed for collation support
+    val collation = CollationFactory.fetchCollation(collationId)
+    collationSupportLevel match {
+      case CollationSupportLevel.SUPPORT_BINARY_ONLY =>
+        if (!collation.isBinaryCollation) {
+          throwUnsupportedCollation(functionName, collation.collationName)
+        }
+      case CollationSupportLevel.SUPPORT_LOWERCASE =>
+        if (!collation.isBinaryCollation && !collation.isLowercaseCollation) {
+          throwUnsupportedCollation(functionName, collation.collationName)
+        }
+      case CollationSupportLevel.SUPPORT_ALL_COLLATIONS => // No additional checks needed
+      case _ => throw new IllegalArgumentException("Invalid collation support level.")
```

Comment:
Please, use `SparkIllegalArgumentException` if it is user-facing error, or switch to an `SparkException.internalError`.